x-pack/plugin/ccr/src/main/java/org/elasticsearch/xpack/ccr/repository/CcrRepository.java

```diff
@@ -413,22 +412,25 @@ public void restoreShard(
                 CcrRetentionLeases.RETENTION_LEASE_RENEW_INTERVAL_SETTING.get(store.indexSettings().getNodeSettings()),
                 Ccr.CCR_THREAD_POOL_NAME
             );
+            // Some tests depend on closing session before cancelling retention lease renewal
             toClose.add(() -> {
                 logger.trace("{} canceling background renewal of retention lease [{}] at the end of restore", shardId, retentionLeaseId);
                 renewable.cancel();
             });
             // TODO: There should be some local timeout. And if the remote cluster returns an unknown session
             // response, we should be able to retry by creating a new session.
-            final RestoreSession restoreSession = openSession(metadata.name(), remoteClient, leaderShardId, shardId, recoveryState);
-            toClose.addFirst(restoreSession); // Some tests depend on closing session before cancelling retention lease renewal
-            restoreSession.restoreFiles(store, restoreListener.delegateFailureAndWrap((l, v) -> {
-                logger.trace("[{}] completed CCR restore", shardId);
-                updateMappings(remoteClient, leaderIndex, restoreSession.mappingVersion, client, shardId.getIndex());
-                l.onResponse(null);
-            }));
-        } catch (Exception e) {
-            restoreListener.onFailure(e);
-        }
+            ActionListener<RestoreSession> sessionListener = restoreListener1.delegateFailureAndWrap((l1, restoreSession) -> {
+                restoreSession.restoreFiles(
+                    store,
+                    l1.delegateResponse((listener1, e) -> restoreSession.close(listener1)).delegateFailureAndWrap((l2, v) -> {
```

Comment:
hmmm... too many listeners! Is this correct? My thinking was that besides closing at the end of the happy path (and before `toClose` is closed since apparently some tests need this order), we'd need to close if there is an exception, therefore the `delegateResponse`. Is there a better way of doing this?

Reply:
I've pushed https://github.com/elastic/elasticsearch/pull/98035/commits/b17594c1152e339fc7edac077758a6815e8dce00.